### DISCUSSION TITLE
MacOS M1 ImGui_ImplOpenGL3_CreateDeviceObjects: failed to compile vertex shader! With GLSL: #version 130

### DISCUSSION THREAD

**Discussion by reformstudios** [ocornut/imgui#6460] [New Users Build/Link/Run issues ONLY!]
Hi there, I'm new to c++ and imgui. Thanks in advance for any help you can provide! 

I managed to get this working on windows, but I'm having trouble on osx (M1, VScode, cmake).

- Can anyone confirm that imgui can work on M1 macos? 
- How can I configure glfWindowHint so it can support imgui on macos? 

The example repo I'm using is here : https://github.com/Bktero/HelloWorldWithDearImGui

I update cmakelists.txt to latest versions of glew(glew-cmake-2.2.0) and glfw (3.3.8). (see bottom of post for code). This resolved an issue that prevented the opengl window from opening.

I've installed glew and glfw with brew.

I've updated main.cpp with glfWindowHint calls to support macos : 

"""

    const char* glsl_version = "#version 130";
    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
    glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
    glfwWindowHint(GLFW_OPENGL_PROFILE,GLFW_OPENGL_CORE_PROFILE);
    glfwWindowHint(GLFW_RESIZABLE, GL_TRUE);

"""

Env info : 
oepngl shader version: 4.1
4.1 Metal - 83

When the app runs, I get a blank opengl window with the following output : 

"""
 
    opengl shader version: 4.1
    4.1 Metal - 83
    ERROR: ImGui_ImplOpenGL3_CreateDeviceObjects: failed to compile vertex shader! With GLSL: #version 130
    
    ERROR: 0:1: '' :  version '130' is not supported
    ERROR: 0:2: '' :  #version required and missing.
    
    ERROR: ImGui_ImplOpenGL3_CreateDeviceObjects: failed to compile fragment shader! With GLSL: #version 130
    
    ERROR: 0:1: '' :  version '130' is not supported
    ERROR: 0:2: '' :  #version required and missing.
    
    ERROR: ImGui_ImplOpenGL3_CreateDeviceObjects: failed to link shader program! With GLSL #version 130
    
    ERROR: One or more attached shaders not successfully compiled

"""

# CMakeLists.txt

"""

    cmake_minimum_required(VERSION 3.17)
    project(HelloWorldWithDearImGui)
    
    #-----------------------------------------------------------------------------------------------------------------------
    # Download dependencies for GitHub
    include(FetchContent)
    
    FetchContent_Declare(
            glew
            GIT_REPOSITORY https://github.com/Perlmint/glew-cmake.git
            GIT_TAG glew-cmake-2.2.0
    )
    
    FetchContent_Declare(
            glfw
            GIT_REPOSITORY https://github.com/glfw/glfw
            GIT_TAG 3.3.8
    )
    
    FetchContent_Declare(
            imgui
            GIT_REPOSITORY https://github.com/ocornut/imgui
            GIT_TAG v1.89.5
    )
    
    FetchContent_MakeAvailable(glew glfw imgui)
    
    # FetchContent_Declare() normally supports EXCLUDE_FROM_ALL but apparently there is a bug
    # See https://gitlab.kitware.com/cmake/cmake/-/issues/20167
    # As a workaround, we exclude the directories manually
    set_property(DIRECTORY ${glew_SOURCE_DIR} PROPERTY EXCLUDE_FROM_ALL YES)
    set_property(DIRECTORY ${glfw_SOURCE_DIR} PROPERTY EXCLUDE_FROM_ALL YES)
    
    #-----------------------------------------------------------------------------------------------------------------------
    # Configure C++ version
    
    set(CMAKE_CXX_STANDARD 20)
    set(CMAKE_CXX_STANDARD_REQUIRED ON)
    set(CMAKE_CXX_EXTENSIONS OFF)
    
    #-----------------------------------------------------------------------------------------------------------------------
    # Create executable
    set(EXECUTABLE_NAME ${PROJECT_NAME})
    
    add_executable(${EXECUTABLE_NAME}
            main.cpp
    
            ${imgui_SOURCE_DIR}/imgui.cpp
            ${imgui_SOURCE_DIR}/imgui.h
            ${imgui_SOURCE_DIR}/imgui_demo.cpp
            ${imgui_SOURCE_DIR}/imgui_draw.cpp
            ${imgui_SOURCE_DIR}/imgui_internal.h
            ${imgui_SOURCE_DIR}/imgui_tables.cpp
            ${imgui_SOURCE_DIR}/imgui_widgets.cpp
    
            ${imgui_SOURCE_DIR}/backends/imgui_impl_glfw.cpp
            ${imgui_SOURCE_DIR}/backends/imgui_impl_opengl3.cpp
    
            )
    
    # Add options for warning
    if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
        target_compile_options(${EXECUTABLE_NAME}
                PRIVATE
                -Wall -Wextra -pedantic
                )
    endif ()
    
    # Add a define to use the static version of Glew
    target_compile_definitions(${EXECUTABLE_NAME}
            PRIVATE
            GLEW_STATIC
            )
    
    # Set the include path
    target_include_directories(${EXECUTABLE_NAME}
            PRIVATE
            ${imgui_SOURCE_DIR}/
            )
    
    # Link against the appropriate libraries
    target_link_libraries(${EXECUTABLE_NAME}
            PRIVATE
            glfw
            libglew_static
            $<$<CXX_COMPILER_ID:GNU>:opengl32.a>
            $<$<CXX_COMPILER_ID:MSVC>:opengl32.lib>
            )

"""

# main.cpp

"""

    #include <iostream>   
    #include "GL/glew.h"
    #include "GLFW/glfw3.h"
    #include "imgui.h"
    #include "backends/imgui_impl_glfw.h"
    #include "backends/imgui_impl_opengl3.h"
    
    static void helloWorld() {
	    ImGui::Begin("My DearImGui Window");
	    ImGui::Text("hello, world");
	    ImGui::End();
    }
    
    int main() {
    
	    // Setup window
	    glfwSetErrorCallback([](int error, const char* description) {
		    fprintf(stderr, "Glfw Error %d: %s\n", error, description);
	    });
    
	    if (!glfwInit()) {
        std::cout << "glfw init err" << std::endl;
        return -1;
      }
    
      // Define version and compatibility settings
      const char* glsl_version = "#version 130";
      glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
      glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
      glfwWindowHint(GLFW_OPENGL_PROFILE,GLFW_OPENGL_CORE_PROFILE);
      glfwWindowHint(GLFW_RESIZABLE, GL_TRUE);
      #ifdef __APPLE__
        std::cout << "I'm apple machine" << std::endl;
        glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
      #endif
      // Create OpenGL window and context
      GLFWwindow* window = glfwCreateWindow(800, 600, "OpenGL", NULL, NULL);
      unsigned int major = glfwGetWindowAttrib(window, GLFW_CONTEXT_VERSION_MAJOR);
      unsigned int minor = glfwGetWindowAttrib(window, GLFW_CONTEXT_VERSION_MINOR);
      std::cout << "oepngl shader version: " << major << "." << minor << std::endl;
      glfwMakeContextCurrent(window);
    
      // Check for window creation failure
      if (!window)
      {
          // Terminate GLFW
          glfwTerminate();
          return 0;
      }
    
      // Initialize GLEW
      // glewExperimental = GL_TRUE; glewInit();
      if (glewInit() != GLEW_OK) {
        std::cout << "Error ! " << std::endl;
      }
      std::cout << glGetString(GL_VERSION) << std::endl;
    
	    // GL 3.0 + GLSL 130 - COMMENTED OUT DUE TO CRASH ON ARM MACOS
	    // const char* glsl_version = "#version 130";
	    // glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
	    // glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 0);
    
	    // Create window with graphics context
	    // GLFWwindow* window = glfwCreateWindow(800, 600, "My GLFW Window", nullptr, nullptr);
	    // glfwMakeContextCurrent(window);
	    glfwSwapInterval(1); // Enable vsync
    
	    // Initialize OpenGL loader
	    // glewInit();
    
	    // Setup Dear ImGui context
	    IMGUI_CHECKVERSION();
	    ImGui::CreateContext();
    
	    // Setup Dear ImGui style
	    ImGui::StyleColorsDark();
    
	    // Setup Platform/Renderer bindings
	    ImGui_ImplGlfw_InitForOpenGL(window, true);
	    ImGui_ImplOpenGL3_Init(glsl_version);
    
	    // Main loop
	    int display_w, display_h;
	    glfwGetFramebufferSize(window, &display_w, &display_h);
    
	    while (!glfwWindowShouldClose(window)) {
		    // Poll and handle events (inputs, window resize, etc.)
		    glfwPollEvents();
    
		    // Start the Dear ImGui frame
		    ImGui_ImplOpenGL3_NewFrame();
		    // ImGui_ImplGlfw_NewFrame();
		    // ImGui::NewFrame();
    
		    // Run our Dear ImGui application
		    // helloWorld();
    
		    // Rendering
		    // ImGui::Render();
		    glfwGetFramebufferSize(window, &display_w, &display_h);
		    glViewport(0, 0, display_w, display_h);
		    glClearColor(1.0f, 1.0f, 1.0f, 1.0f);
		    glClear(GL_COLOR_BUFFER_BIT);
		    // ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());
		    glfwSwapBuffers(window);
	    }
    
	    // Cleanup
	    ImGui_ImplOpenGL3_Shutdown();
	    ImGui_ImplGlfw_Shutdown();
	    ImGui::DestroyContext();
    
	    glfwDestroyWindow(window);
	    glfwTerminate();
    }

"""

**Comment by reformstudios** [comment]
I tried copying the example_glfw_opengl3 example code in to my main.cpp and after adding "backend/" to the includes, it worked! 
So, Some of the builderplate in the example obviously resolves this issue. 
You can close this ticket, but I'm curious to understand what was wrong with the previous code. 
Also, I'm trying to better understand CMakeLists.txt, so the interesting thing here is why I needed to add "backend/" to my includes. 
Is it possible to update my CMakeLists.txt so that these includes :

#include "backends/imgui_impl_glfw.h"
#include "backends/imgui_impl_opengl3.h"

Can remain as in the example, eg : 

#include "imgui_impl_glfw.h"
#include "imgui_impl_opengl3.h"

Thanks again!

**Comment by reformstudios** [comment]
Ok, Confirmed the offending line was :-
"""
    const char* glsl_version = "#version 130";
"""
which should have been 
"""
    const char* glsl_version = "#version 150";
"""

**Reply by xbz-24** [reply]
thanks, this fixed it for me
